src/idiomatic/leveraging-the-type-system/raii/drop_limitations.md

```diff
@@ -0,0 +1,84 @@
+# The limitations of `Drop`
+
+While `Drop` works well for cases like synchronization primitives, its use
+becomes more questionable when dealing with I/O or unsafe resources.
+
+```rust
+use std::fs::File;
+use std::io::{self, Write};
+
+fn write_log() -> io::Result<()> {
+    let mut file = File::create("log.txt")?;
+    // ^ ownership of the (OS) file handle starts here
+
+    writeln!(file, "Logging a message...")?;
+    Ok(())
+} // file handle goes out of scope here
+```
+
+<details>
+
+- In the earlier example, our `File` resource owns a file handle provided by the
+  operating system.
+
+  [As stated in the documentation](https://doc.rust-lang.org/std/fs/struct.File.html):
+
+  > Files are automatically closed when they go out of scope. Errors detected on
+  > closing are ignored by the implementation of Drop.
+
+- This highlights a key limitation of the `Drop` trait: it cannot propagate
+  errors to the caller. In other words, fallible cleanup logic cannot be handled
+  by the code using the `File`.
+
+  This becomes clear when looking at the
+  [definition of the `Drop` trait](https://doc.rust-lang.org/std/ops/trait.Drop.html):
+
+  ```rust
+  trait Drop {
+      fn drop(&mut self);
+  }
+  ```
+
+  Since `drop` does not return a `Result`, any error that occurs during cleanup
+  cannot be surfaced or recovered from. This is by design: `drop` is invoked
+  automatically when a value is popped off the stack during unwinding, leaving
+  no opportunity for error handling.
+
+- One workaround is to panic inside `drop` when a failure occurs. However, this
+  is risky—if a panic happens while the stack is already unwinding, the program
+  will abort immediately, and remaining resources will not be cleaned up.
+
+  While panicking in `drop` can serve certain purposes (see
+  [the next chapter on "drop bombs"](./drop_bomb.md)), it should be used
+  sparingly and with full awareness of the consequences.
+
+- Another drawback of `drop` is that its execution is implicit and
+  non-deterministic in terms of timing. You cannot control _when_ a value is
+  dropped. And in fact as discussed in previous slide it might never even run at
+  all, leaving the external resource in an undefined state.
+
+  This matters particularly for I/O: normally you might set a timeout on
+  blocking operations, but when I/O occurs in a `drop` implementation, you have
+  no way to enforce such constraints.
+
+  Returning to the `File` example: if the file handle hangs during close (e.g.,
+  due to OS-level buffering or locking), the drop operation could block
+  indefinitely. Since the call to `drop` happens implicitly and outside your
+  control, there's no way to apply a timeout or fallback mechanism.
```

Comment:
I don't understand this point. One would apply a timeout in the same way as usual. Where we have a limitation is that `drop()` can't communicate with the caller: it can neither receive an argument (e.g., a timeout value) or communicate back that the timeout happened (the whole discussion about failures in drop). But certainly `drop()` is free to call a low-level I/O function with a timeout value - hence I'm confused about what this paragraph is trying to communicate.

The only limitation of drop I can think of that can make I/O difficult is that drop is not async. There's the [unstable AsyncDrop](https://doc.rust-lang.org/std/future/trait.AsyncDrop.html) for that.

Reply:
I meant that as a user of the API you have no control over it.

As in:

- when you can `Foo::close` you can wrap around that logic however you want. Not just error handling, but also impose a timeout, retries, etc...
- that is not something you can do "as the user of the API" for something that happens invisible in the background by the creator of the API in Drop.